contracts/contracts/harvest/BaseHarvester.sol

```diff
@@ -99,87 +133,208 @@ abstract contract BaseHarvester is Governable {
-            oldUniswapAddress != _uniswapV2CompatibleAddr
+            oldRouterAddress != address(0) &&
+            oldRouterAddress != newRouterAddress
         ) {
-            token.safeApprove(oldUniswapAddress, 0);
+            token.safeApprove(oldRouterAddress, 0);
+        }
+
+        // Give SwapRouter infinite approval when needed
+        if (oldRouterAddress != newRouterAddress) {
+            token.safeApprove(newRouterAddress, 0);
+            token.safeApprove(newRouterAddress, type(uint256).max);
         }
 
-        // Give Uniswap infinite approval when needed
-        if (oldUniswapAddress != _uniswapV2CompatibleAddr) {
-            token.safeApprove(_uniswapV2CompatibleAddr, 0);
-            token.safeApprove(_uniswapV2CompatibleAddr, type(uint256).max);
+        SwapPlatform _platform = tokenConfig.platform;
+        if (_platform == SwapPlatform.UniswapV2Compatible) {
+            uniswapV2Path[_tokenAddress] = _decodeUniswapV2Path(
+                swapData,
+                _tokenAddress
+            );
+        } else if (_platform == SwapPlatform.UniswapV3) {
+            uniswapV3Path[_tokenAddress] = _decodeUniswapV3Path(
+                swapData,
+                _tokenAddress
+            );
+        } else if (_platform == SwapPlatform.Balancer) {
+            balancerPoolId[_tokenAddress] = _decodeBalancerPoolId(swapData);
+        } else if (_platform == SwapPlatform.Curve) {
+            curvePoolIndices[_tokenAddress] = _decodeCurvePoolIndices(
+                swapData,
+                newRouterAddress,
+                _tokenAddress
+            );
+        } else {
+            revert("Unknown SwapPlatfrom");
         }
 
         emit RewardTokenConfigUpdated(
             _tokenAddress,
-            _allowedSlippageBps,
-            _harvestRewardBps,
-            _uniswapV2CompatibleAddr,
-            _liquidationLimit,
-            _doSwapRewardToken
+            tokenConfig.allowedSlippageBps,
+            tokenConfig.harvestRewardBps,
+            _platform,
+            newRouterAddress,
+            swapData,
+            tokenConfig.liquidationLimit,
+            tokenConfig.doSwapRewardToken
+        );
+    }
+
+    /**
+     * @dev Decodes the data passed into Uniswap V2 path and validates
+     *      it to make sure the path is for `token` to `baseToken`
+     *
+     * @param data Ecnoded data passed to the `setRewardTokenConfig`
+     * @param token The address of the reward token
+     * @return path The validated Uniswap V2 path
+     */
+    function _decodeUniswapV2Path(bytes calldata data, address token)
+        internal
+        view
+        returns (address[] memory path)
+    {
+        (path) = abi.decode(data, (address[]));
+
+        // Do some validation
+        require(
+            path.length >= 2 &&
+                path[0] == token &&
+                path[path.length - 1] == baseTokenAddress,
+            "Invalid Uniswap V2 path"
```

Comment:
you could do custom errors here for each of the 3 conditions and pass the incorrect value to an error as a parameter